### ISSUE TITLE
Improve correspondence between packages and plugins

### ISSUE THREAD

**Issue by DragaDoncila** [napari/napari#3035]
## 🚀 Feature
A napari plugin is a defined entry point in a Python package. 
The napari plugin install dialog displays entry points as individual plugins. 
Multiple plugins can exist inside a single package. 
However, it's the python package containing the plugins that can be installed/uninstalled by the user.

This could be confusing for users for a number of reasons:
- multiple packages could declare identically named entry points, and only the entry point name is shown in the hook sorter
- there is no enforced correspondence between package names and entry point names so the user might have trouble identifying which package a plugin belongs to and how it came to exist in their environment
- when a package provides multiple plugins, **all** plugins are uninstalled when the user clicks uninstall on one of them (see video)

https://user-images.githubusercontent.com/17995243/125367463-85e99780-e3bb-11eb-9e2a-3d69a3ba388b.mp4

My suggestion is that entry points are somehow visually grouped under their respective package to clarify the relationship between them, and that both package and entry point are displayed in the plugin sorter to avoid confusion. I'm not sure whether we should also be showing the package in the widget menu? It's going to make it so wordy, but I do think the potential for confusion for users is too real to ignore, particularly as users start installing from the napari-hub where there is no distinction between package and plugin.

I'd love to get people's thoughts on how we can clarify these relationships in the viewer. @tlambert03 @sofroniewn @jni ? I think this also highlights the need for further discussion around a best practice guide for developers for the hierarchical organisation of their packages, and perhaps an informative diagram describing the relationship between package - plugin (entry point) - hook specification - hook implementation.

**Comment by DragaDoncila** [comment]
Pinging @neuromusic and @liaprins-czi as they may wish to track progress on this issue.

**Comment by tlambert03** [comment]
Totally agree @DragaDoncila ... I've been concerned about this too and was thinking of raising a similar issue.

My immediate concern isn't so much that a user gets two plugins for one package... or how they are presented in the plugin installer (though, I do see the potential confusion there).  I _am_ concerned a lot however about the fact that multiple packages can declare the same entry_point.

This was more or less inherited from pluggy... but it basically just needs to stop somehow, and soon.  At the same time, we need a way to allow a plugin to have a "descriptive" name that is independent of their package name (so that something like `napari -w omero` still works, even though the package isn't called omero).

Package name is the one thing that we have that is guaranteed not to clash (at least on PyPI ... all bets off for conda, which is also really important).

no immediate solutions here.  but a hearty agreement that this is a high priority issue

**Comment by sofroniewn** [comment]
> I am concerned a lot however about the fact that multiple packages can declare the same entry_point.
> This was more or less inherited from pluggy... but it basically just needs to stop somehow, and soon. 

Yeah, that's not ideal

> At the same time, we need a way to allow a plugin to have a "descriptive" name that is independent of their package name (so that something like napari -w omero still works, even though the package isn't called omero).

Hmm yeah, not sure what to do here either

> no immediate solutions here. but a hearty agreement that this is a high priority issue

Yup, that's how I feel too!!

**Comment by sofroniewn** [comment]
> My suggestion is that entry points are somehow visually grouped under their respective package to clarify the relationship between them, and that both package and entry point are displayed in the plugin sorter to avoid confusion. 

We can also start with some improvements like this too, that might make it easier for users to understand grouping

**Comment by liaprins-czi** [comment]
> My suggestion is that entry points are somehow visually grouped under their respective package to clarify the relationship between them, and that both package and entry point are displayed in the plugin sorter to avoid confusion.

Yes I thin kthis is very intuitive and natural first step. Also makes it more clear about what happens + how to uninstall

**Comment by neuromusic** [comment]
thanks for adding this issue, @DragaDoncila!

my biggest concern is that we use clear and consistent nomenclature within and across the napari viewer, the napari hub, and other materials.

for v0 of the hub, we are referring to the python package as the "plugin". this is consistent with... 

- the napari plugin installation GUI, which lists python packages under "Available Plugins"
- parts of the documentation for creating a plugin, which lists `napari-dv` and `napari-aicsimageio` as example plugins
- with how the community tends to refer to the package and plugins as being one-and-the-same (which is often true, since the vast majority of packages only implement a single entry_point)

however, this is inconsistent with...

- the napari plugin installation GUI, which lists the entry_points under "Installed Plugins"
- parts of the documentation for creating a plugin which clarifies that the "plugin name" is defined in the entry_point (thus a single package can implement >1 plugin)
- my understanding of @tlambert03 's intent with the plugin architecture & how napari developers are careful to distinguish between the package and the plugin

`napari-aicsimageio` is a great example of where this ambiguity around nomenclature pops up across the two projects and how the plugin dev describes its use: 

- the napari hub lists `napari-aicsimageio` as a "plugin"
- on its listing, @JacksonMaxfield tells his users that the `napari-aicsimageio` plugin has two "variants" (defined with two entry_points) & describes their different uses
- if the user clicks "install", we provide instructions for installing the `napari-aicsimageio` "plugin"
- in napari, `napari-aicsimageio` is listed as a single "Available Plugin"
- only after install are the two entry points are listed as separate plugins

if we agree that "a plugin is defined by the entry_point" and that a package can contain more than one plugin, then we need to refer to the Python package as something else... a "package"? a "community package"? a "napari package"? idk. but we stop calling the package a "plugin", even if there is only one defined. the package contains the plugin or defines the plugin or makes the plugin available.

in this world...
- "Available Plugins" becomes "Available Packages"
- we make the UX improvements to the install/remove process above
- on the napari hub, we overhaul our language and paths
- as we look into inspecting plugins and adding more metadata to the hub about e.g. which hook specs a package implements, we also consider surfacing info about the plugin (e.g. add "number of plugins" as metadata or list the names of the plugins that a package contains and make the plugin names searchable)

Alternatively, if we want to follow the "a plugin is defined by the package" approach, then the entry_point needs to become something else for end users... a "variant"? a "module"? a "sub-plugin"? an "entry point"? I don't think this is unreasonable, but I also don't think its consistent with the intent of the plugin architecture

If we can agree on consistent nomenclature for users, I'm happy to make sure that the napari hub follows the lead of the napari viewer

**Comment by sofroniewn** [comment]
Thanks for the detailed post here on nomenclature @neuromusic - maybe a topic for an upcoming community meeting discussion, just to get it on folks minds. Curious what @jni thinks here too

**Comment by sofroniewn** [comment]
@goanpeca see discussion here, might be relevant for "multiplugin" install case

**Comment by neuromusic** [comment]
Following up on Thursday's call... 

In our discussion there, @sofroniewn & @jni expressed support for "package=plugin".

This would mean that @DragaDoncila's setup for this issue needs some revision. Specifically,

> A napari plugin is a defined entry point in a Python package.

Would become 

> A napari plugin is a ~defined entry point in a~ Python package.

**Comment by goanpeca** [comment]
This still leaves the plugin installation dialog in a weird state, since the thing that you see on the top list is not the same that you see on the bottom list. For the sake of cleaning things out I would suggest that the plugin installation dialog dealt only with packages/plugins, so that the top list would only list installed packages, and the bottom list would only list installable packages. (same language)

I would also advocate to removing the plugin sorter (also mixing terms) from the plugin/package install dialog. So that that dialog only deals with python (eventually conda) packages not with entry points. then the plugin sorter would need a different name, since it is talking about different entries that a plugin/package may provide.

This would also allow to cleanup the usage of the `PluginListItemWidget` that merges both behaviors (it can be a package or an entry point).

**Comment by tlambert03** [comment]
> the top list would only list installed packages, and the bottom list would only list installable packages. (same language)
>I would also advocate to removing the plugin sorter (also mixing terms) from the plugin/package install dialog.

both fine by me

**Comment by sofroniewn** [comment]
>  For the sake of cleaning things out I would suggest that the plugin installation dialog dealt only with packages/plugins, so that the top list would only list installed packages, and the bottom list would only list installable packages.

That makes sense to me.

> I would also advocate to removing the plugin sorter (also mixing terms) from the plugin/package install dialog. So that that dialog only deals with python packages not entry points. then the plugin sorter would need a different name, since it is talking about different entries that a plugin/package may provide.

I'd also be ok with removing it there as it is now in the preferences and i don't use it in the plugin installer. I'm ok with sticking with the "plugin sorter" name a little longer as it is in a big picture sense. We could also call it a "hook sorter" or something, but that might just be more confusing

**Comment by goanpeca** [comment]
>"plugin sorter" name a little longer as it is in a big picture sense

Sure, we can also add some text on the plugin dialog in the lines of:

"A napari plugin may provide several hooks, the priority of how these might be used can be set in the plugins preferences dialog"

And on the preferences dialog we might have a text that clarifies the terms.

**Comment by tlambert03** [comment]
I think we should also try to drop support for the multiple entry point thing. It's not really necessary.  I think aicsimageio might be the only package doing it (easy enough to check) so we could help transition anyone away from it

**Comment by goanpeca** [comment]
>I think we should also try to drop support for the multiple entry point thing. It's not really necessary.

Sounds good.

**Comment by goanpeca** [comment]
@tlambert03 what do you think of having then a single list (single search box) sorted by installed/uninstalled and name?

**Comment by tlambert03** [comment]
I like it!  In addition to just sorting though, one should be able to show only installed or only available things (maybe two checkboxes?)

**Comment by goanpeca** [comment]
>I like it! In addition to just sorting though, one should be able to show only installed or only available things (maybe two checkboxes?)

Maybe. I think we need @liaprins-czi help :)

**Comment by goanpeca** [comment]
See https://github.com/napari/napari/pull/3071

**Comment by sofroniewn** [comment]
> > I think we should also try to drop support for the multiple entry point thing. It's not really necessary.
>
> Sounds good.

Good by me too.

> Maybe. I think we need @liaprins-czi help :)

😄  @liaprins-czi we should get something around the plugin installer in the design queue for the future - this can be a good test case for ideas around a design intake process

**Comment by sofroniewn** [comment]
Now that #3069 and #3071 have merged can we close this or is there still more we want to do here? We can also open up a new issue for the additional work if it is different in scope

**Comment by DragaDoncila** [comment]
@sofroniewn I think final issue in regards to this would be a warning or an error when two plugins try to register the same entry point. Even with individual entry points per package, we have no namespace for the entry points. The result is that one plugin completely shadows the other if they have identically named entry points.

Longer term I know @tlambert03 is looking at completely revamping the way developers specify functionality provided by plugins, but in the meantime it's probably good to handle this case specifically?

**Comment by tlambert03** [comment]
yeah, totally agree @DragaDoncila.  As I [mentioned earlier](https://github.com/napari/napari/issues/3035#issuecomment-878674473), I see the user confusion caused by the gui (fixed in those recent PRs) to be a relatively minor problem compared to the potential for namespace clashes.

> a warning or an error when two plugins try to register the same entry point.

@DragaDoncila, if you'd like to submit a PR for this, we'd just need to emit a warning on this line:
https://github.com/napari/napari-plugin-engine/blob/51061cc8da30d9ab05c7ffa56659c141f015cb93/napari_plugin_engine/manager.py#L239-L240
if `self.is_registered(name)` is True, but the module returned by `self.plugins[name]` doesn't match `mod_name`

**Comment by DragaDoncila** [comment]
@tlambert03 yep I was planning on doing so later today!

**Comment by tlambert03** [comment]
**edit**: actually... you could even just go ahead and rename it to `plugin-2` or something, widgets and stuff would still show up in the menu

**Comment by DragaDoncila** [comment]
Yeah I was thinking about renaming, probably still with a warning to the user though

**Comment by DragaDoncila** [comment]
I think we can now close this issue, after merging [#32](https://github.com/napari/napari-plugin-engine/pull/32) in `napari-plugin-engine`
